File: server/twitter/twitter.go
package twitter

import (
	"bytes"
	"encoding/json"
	"fmt"
	twit "github.com/dghubble/go-twitter/twitter"
	"github.com/dghubble/oauth1"
	"log"
	"server/conf"
	"server/db"
	"server/models"
	"sync"
)

type TwitterContext struct {
	database      *db.DBContext
	config        conf.Twitter
	twitterClient *twit.Client
	allowedUserIDsMutex sync.RWMutex
	allowedUserIDs map[string]bool
}

func NewTwitterContext(database *db.DBContext, config conf.Twitter) (*TwitterContext, error) {
	oauthConfig := oauth1.NewConfig(config.ConsumerKey, config.ConsumerSecret)
	token := oauth1.NewToken(config.AccessToken, config.AccessTokenSecret)
	twitterClient := twit.NewClient(oauthConfig.Client(oauth1.NoContext, token))

	twitterContext := &TwitterContext{
		database:       database,
		config:         config,
		twitterClient:  twitterClient,
		allowedUserIDs: make(map[string]bool),
	}

	err := twitterContext.updatedAllowedUserIDs()
	if err != nil {
		return nil, fmt.Errorf("twitter NewTwitterContext initial fetch of allowed user IDs failed: %v", err)
	}

	return twitterContext, nil
}

func (ctx *TwitterContext) updatedAllowedUserIDs() error {
	allowedUserIDsSlice, err := ctx.database.GetWhitelistTwitterIDs()
	if err != nil {
		return fmt.Errorf("twitter updateAllowedUserIDs getting allowed user IDs from database failed: %v", err)
	}

	allowedUserIDsMap := make(map[string]bool)
	for _, userID := range allowedUserIDsSlice {
		allowedUserIDsMap[userID] = true
	}

	ctx.allowedUserIDsMutex.Lock()
	ctx.allowedUserIDs = allowedUserIDsMap
	ctx.allowedUserIDsMutex.Unlock()
	return nil
}

func (ctx *TwitterContext) onTweet(t *twit.Tweet) {
	jsonBody := &bytes.Buffer{}
	err := json.NewEncoder(jsonBody).Encode(t)
	if err != nil {
		log.Printf("twitter onTweet encoding tweet body to JSON failed: %v", err)
	}

	reviewRequired := true
	ctx.allowedUserIDsMutex.RLock()
	if _, exists := ctx.allowedUserIDs[t.User.IDStr]; exists {
		reviewRequired = false
	}
	ctx.allowedUserIDsMutex.RUnlock()

	time, _ := t.CreatedAtTime()

	err = ctx.database.SaveTweet(models.Tweet{
		TweetID:        t.IDStr,
		AuthorID:       t.User.IDStr,
		Time:           time,
		JSONBody:       jsonBody.Bytes(),
		ReviewRequired: reviewRequired,
	})

	if err != nil {
		log.Printf("twitter onTweet saving tweet to database failed: %v", err)
	}
}

func (ctx *TwitterContext) StartStream() error {
	params := &twit.StreamFilterParams{
		Track: []string{ctx.config.TextToTrack},
	}

	demux := twit.NewSwitchDemux()
	demux.Tweet = ctx.onTweet

	stream, err := ctx.twitterClient.Streams.Filter(params)
	if err != nil {
		return fmt.Errorf("twitter StartStream failed: %v", err)
	}

	// Receive messages until app quits
	go demux.HandleChan(stream.Messages)
	return nil
}

File: server_node/src/lib/twitter.js
const knex = require('knex');
const twit = require('twit');

class Twitter {
  /**
   * @param {*} config
   * @param {knex} db
   */
  constructor(config, db) {
    this.client = new twit(config);
    this.config = config;
    this.db = db;
    this.whitelistedUserIds = new Set();
    this.mutedUserIds = new Set();
  }

  async updateMutedUserIds() {
    const mutedUserIds = await this.db
      .select('twitter_user_id')
      .from('muted_tweeters');

    this.mutedUserIds = new Set(mutedUserIds);
  }

  async updateWhitelistedUserIds() {
    const whitelistedUserIds = await this.db
      .select('twitter_user_id')
      .from('whitelisted_tweeters');

    this.whitelistedUserIds = new Set(whitelistedUserIds);
  }

  async startStream() {
    await Promise.all([
      this.updateMutedUserIds(),
      this.updateWhitelistedUserIds()
    ]);

    const stream = this.client.stream('statuses/filter', {
      track: this.config.textToTrack
    });

    stream.on('tweet', async tweet => {
      if (this.mutedUserIds.has(tweet.user.id_str)) {
        return;
      }

      try {
        await this.db('tweets').insert({
          tweet_id: tweet.id_str,
          timestamp: new Date(tweet.created_at),
          original_tweet_json: JSON.stringify(tweet),
          author_id: tweet.user.id_str,
          review_required: !this.whitelistedUserIds.has(tweet.user.id_str)
        });
      } catch (err) {
        console.error('Failed to save tweet to database', err);
      }
    });

    stream.on('error', err => {
      console.error('Twitter stream error', err);
    });
  }
}

module.exports = { Twitter };
